cpp/src/arrow/type_fwd.h

```diff
@@ -351,39 +358,21 @@ std::shared_ptr<DataType> ARROW_EXPORT time64(TimeUnit::type unit);
 std::shared_ptr<DataType> ARROW_EXPORT
 struct_(const std::vector<std::shared_ptr<Field>>& fields);
 
-/// \brief Create a UnionType instance
-std::shared_ptr<DataType> ARROW_EXPORT
-union_(const std::vector<std::shared_ptr<Field>>& child_fields,
```

Comment:
I think we should keep this factory function, for compatibility and for convenience. `UnionMode` still exists, right?

Reply:
UnionMode still exists but I've already factored out usage of `union_`. If we need it later it'll be straightforward to add

Reply:
This is more of a "don't break third-party code" concern.

Reply:
I'm quickly restoring backward compatibility functions